### ISSUE TITLE
ITenantProvider.GetTenantAsync throws exception for common case

### ISSUE THREAD

**Issue by idg10** [corvus-dotnet/Corvus.Tenancy#80]
Currently, the `ITenantProvider.GetTenantAsync` method throws a `TenantNotModifiedException` if the stored tenant's etag matches the etag passed in.

This is likely to be the most common outcome—tenants are fetched all the time but are not modified very often. So not only are we using an exception to signal a fairly common result (which violates the .NET Framework design guidelines), we're doing so in the most common case!

It would arguably be more sensible for the returned task to produced `null` if the etag matches. We're not currently using a `null` result to signify anything else. (We throw a different exception in the case where there's no tenant with the specified id.)

**Comment by mwadams** [comment]
Certainly less painful! I agree this would be a good (basically internal) optimisation.

**Comment by mwadams** [comment]
Of course, e.g. the Cosmos SDK throws in these circumstances (409 conflict generates an exception)

**Comment by mwadams** [comment]
I've seen these SDKs oscillate between exceptions and return codes so often that I'm unsure what "best" looks like - in this case forcing people to deal with nulls, or forcing them to deal with exceptions.

My instinct is still towards exceptions. However, that requires more work all round - including in the Retry handling.

**Comment by idg10** [comment]
> Of course, e.g. the Cosmos SDK throws in these circumstances (409 conflict generates an exception)

even for a GET? I thought the correct HTTP status code when you perform a GET passing an etag in an `If-None-Match` header was 304 (Not modified).

**Comment by mwadams** [comment]
Sorry - quite correct. 304.

**Comment by mwadams** [comment]
That will still come as an exception though. So the point is still valid!

**Comment by idg10** [comment]
I think an exception on a 409 is reasonable, but it seems bonkers on a 304 in this case. Thats a cache hit. Cache hits are not exceptional! So if that's what the Cosmos SDK is doing in this case I think it's definitely wrong.

**Comment by mwadams** [comment]
https://github.com/corvus-dotnet/Corvus.Tenancy/blob/2e5d3f221fc70637db793c7bb1292da418c12489/Solutions/Corvus.Tenancy.Storage.Azure.Blob/Corvus/Tenancy/Storage/Azure/TenantProviderBlobStore.cs#L332

**Comment by mwadams** [comment]
We have to handle this storage exception in this case.

**Comment by idg10** [comment]
So having reminided myself of how the Azure Storage SDK handles this, this I've concluded that they've done it this way because:

* they have a general model for preconditions (`AccessCondition`)
* in some of the cases, preconditions not being met are genuinely exceptional (e.g., conflict)
* defining two completely separate mechanisms (exceptional and non-exception) for reporting when specified access conditions are not met would probably have caused more problems than it solved
* they didn't want to force people who don't care about preconditions do deal with them by making them part of the return type of every single operation

Given all of that, I can see how throwing in this particular case looked like a better idea than the alternatives.

However, I don't think the same chain of reasoning applies to us. This method has already decided not to use a general-purpose precondition mechanism by taking an `eTag` argument, so since the method signature already specializes on one very particular case—the "is the copy in my local cache up to date?" case—none of the factors listed above that led to the storage SDK deciding to throw for this case apply.

The one argument I can think of in favour of exeptions for us would be that in a nullable world, if you have no interest in the "unless my cached copy is still valid" semantics (i.e., you're not passing an ETag) then it's actually quite annoying for the returned tasks's result type to be nullable. And the advantage of reporting a cache hit via an exception is that it means you can make the result type non-nullable.

However, that could also be addressed by having two separate methods: one that doesn't take an ETag and which returns a task producing a non-nullable `ITenant`, and one that does take an ETag, returning a task that produces an `ITenant?`.
